File: desktop/captions_desktop/vad.py
"""Energy-based VAD (Python port of the PWA's, so both builds endpoint alike).

Adaptive noise floor + on/off hangover. A Silero-VAD upgrade can sit behind the
same ``process()`` interface later.
"""

from __future__ import annotations

import math
from typing import Optional, Sequence


class EnergyVAD:
    def __init__(
        self,
        sample_rate: int = 16000,
        start_ms: float = 120,
        end_ms: float = 600,
        margin: float = 2.5,
    ) -> None:
        self.sample_rate = sample_rate
        self.start_ms = start_ms
        self.end_ms = end_ms
        self.margin = margin
        self.noise_floor = 0.005
        self.speaking = False
        self._speech_ms = 0.0
        self._silence_ms = 0.0

    def process(self, frame: Sequence[float]) -> Optional[str]:
        """Returns 'start', 'end', or None for this frame."""
        rms = _rms(frame)
        dur_ms = (len(frame) / self.sample_rate) * 1000
        threshold = max(self.noise_floor * self.margin, 0.006)
        is_speech = rms > threshold

        if not is_speech:
            self.noise_floor = self.noise_floor * 0.95 + rms * 0.05

        if self.speaking:
            if is_speech:
                self._silence_ms = 0.0
            else:
                self._silence_ms += dur_ms
                if self._silence_ms >= self.end_ms:
                    self.speaking = False
                    self._speech_ms = 0.0
                    return "end"
        elif is_speech:
            self._speech_ms += dur_ms
            if self._speech_ms >= self.start_ms:
                self.speaking = True
                self._silence_ms = 0.0
                return "start"
        else:
            self._speech_ms = 0.0
        return None


def _rms(frame: Sequence[float]) -> float:
    # Fast path for numpy arrays (the live audio frames).
    if hasattr(frame, "size"):
        import numpy as np  # type: ignore

        arr = frame  # type: ignore[assignment]
        if arr.size == 0:  # type: ignore[attr-defined]
            return 0.0
        return float(np.sqrt(np.mean(np.square(arr, dtype=np.float64))))
    if len(frame) == 0:
        return 0.0
    total = 0.0
    for x in frame:
        total += x * x
    return math.sqrt(total / len(frame))

File: packages/pwa/src/audio/vad.ts
/**
 * Lightweight energy-based voice activity detector with an adaptive noise floor
 * and on/off hangover. Good enough to endpoint utterances for live captioning;
 * a Silero-VAD upgrade can slot in behind the same `process()` interface later.
 */
export interface VadOptions {
  /** ms of sustained speech before declaring speech start */
  startMs: number;
  /** ms of sustained silence before declaring speech end */
  endMs: number;
  /** RMS multiple over the noise floor to count as speech */
  margin: number;
}

const DEFAULTS: VadOptions = { startMs: 120, endMs: 600, margin: 2.5 };

export type VadEvent = "start" | "end" | null;

export class EnergyVAD {
  private noiseFloor = 0.005;
  private speaking = false;
  private speechMs = 0;
  private silenceMs = 0;

  constructor(
    private readonly sampleRate = 16000,
    private readonly opts: VadOptions = DEFAULTS,
  ) {}

  get isSpeaking(): boolean {
    return this.speaking;
  }

  process(frame: Float32Array): VadEvent {
    const rms = computeRms(frame);
    const durMs = (frame.length / this.sampleRate) * 1000;
    const threshold = Math.max(this.noiseFloor * this.opts.margin, 0.006);
    const isSpeech = rms > threshold;

    // Adapt the noise floor only while quiet, so it tracks the room not the talker.
    if (!isSpeech) this.noiseFloor = this.noiseFloor * 0.95 + rms * 0.05;

    if (this.speaking) {
      if (isSpeech) {
        this.silenceMs = 0;
      } else {
        this.silenceMs += durMs;
        if (this.silenceMs >= this.opts.endMs) {
          this.speaking = false;
          this.speechMs = 0;
          return "end";
        }
      }
    } else if (isSpeech) {
      this.speechMs += durMs;
      if (this.speechMs >= this.opts.startMs) {
        this.speaking = true;
        this.silenceMs = 0;
        return "start";
      }
    } else {
      this.speechMs = 0;
    }
    return null;
  }
}

function computeRms(frame: Float32Array): number {
  let sum = 0;
  for (let i = 0; i < frame.length; i++) sum += frame[i]! * frame[i]!;
  return Math.sqrt(sum / Math.max(frame.length, 1));
}
